### ISSUE TITLE
Allow to omit return in callback of Query Builder when()

### ISSUE THREAD

**Issue by vlakoff** [laravel/internals#233]
I think https://github.com/laravel/framework/pull/13091 should be reconsidered. Here is an example use case:

``` php
DB::from('foobar')
    ->when($exclude_ids, function ($qb) use ($exclude_ids) {
        return $qb->whereNotIn('id', (array) $exclude_ids);
    })
    ->inRandomOrder()
    ->limit($nb)
    ->get();
```

The "when" part isn't very sexy…, removing the `return` makes it a bit better:

``` php
DB::from('foobar')
    ->when($exclude_ids, function ($qb) use ($exclude_ids) {
        $qb->whereNotIn('id', (array) $exclude_ids);
    })
    ->inRandomOrder()
    ->limit($nb)
    ->get();
```
- The Query Builder instance is mutable, so it doesn't really make sense to return the instance.
- It is backward compatible, if you return anything else from the callback it will be used instead in the chain.
- The change in `when()` implementation is very simple.

**Comment by tomschlick** [comment]
`return` is consistent with the rest of the framework. 

For instance query scopes: https://laravel.com/docs/5.3/eloquent#query-scopes

and almost every method on the collections: https://laravel.com/docs/5.3/collections#method-map

**Comment by vlakoff** [comment]
I think Collection methods are really a different case, where `return` is of course perfectly fine.

About scopes:
* The `return` is less bothering as it is inside simple methods, rather than a full query builder chain.
* In fact, the `return` in scopes `apply()`'s could (should) be omitted, as the return value is not used. See [`Builder::applyScopes()`](https://github.com/laravel/framework/blob/master/src/Illuminate/Database/Eloquent/Builder.php) implementation, and the `@return void` phpdoc for `apply()`.
* For dynamic scopes "scopeFoobar" there is a `$this` fallback similar to what I'm suggesting here. See in `callScope()` implementation: `$result = call_user_func_array($scope, $parameters) ?: $this;`

**Comment by vlakoff** [comment]
See also the code in this recent tweet: https://twitter.com/taylorotwell/status/784407839162851328

```php
$validator = Validator::make(['name' => 'Taylor Otwell'], [
    'name' => [
        Rule::unique('users')->ignore(2)->using(function ($query) {
            $query->whereIn('foo', [1, 2, 3]);
        })
    ],
]);
```

**Comment by vlakoff** [comment]
Implemented in Laravel 5.4.16, by https://github.com/laravel/framework/pull/18422.
